sdl_viewer/src/box_drawer.rs

```diff
@@ -0,0 +1,144 @@
+        }
+
+        let vertex_array = GlVertexArray::new();
+        vertex_array.bind();
+
+        // vertex buffer: define 8 vertices of the box
+        let _buffer_position = GlBuffer::new_array_buffer();
+        _buffer_position.bind();
+        let vertices: [[f32; 3]; 8] = [
+            [-1.0, -1.0,  1.0],   // vertices of front quad
+            [ 1.0, -1.0,  1.0],   //
+            [ 1.0,  1.0,  1.0],   //
+            [-1.0,  1.0,  1.0],   //
+            [-1.0, -1.0, -1.0],   // vertice ofs back quad
+            [ 1.0, -1.0, -1.0],   //
+            [ 1.0,  1.0, -1.0],   //
+            [-1.0,  1.0, -1.0],   //
+        ];
+        unsafe {
+            gl::BufferData(
+                gl::ARRAY_BUFFER,
+                (vertices.len() * 3 * mem::size_of::<f32>()) as GLsizeiptr,
+                mem::transmute(&vertices[0]),
+                gl::STATIC_DRAW,
+            );
+        }
+
+        // define index buffer for 24 edges of the box
+        let _buffer_indices = GlBuffer::new_element_array_buffer();
+        _buffer_indices.bind();
+        let indices: [i32; 24] = [
+            0,1, 1,2, 2,3, 3,0,		// front
+            4,5, 5,6, 6,7, 7,4,		// back
+            1,5, 6,2,				// right
+            4,0, 3,7,				// left
+        ];
+        unsafe {
+            gl::BufferData(
+                gl::ELEMENT_ARRAY_BUFFER,
+                (indices.len() * mem::size_of::<i32>()) as GLsizeiptr,
+                mem::transmute(&indices[0]),
+                gl::STATIC_DRAW,
+            );
+        }
+
+        unsafe {
+            let pos_attr = gl::GetAttribLocation(outline_program.id, c_str!("position"));
+            gl::EnableVertexAttribArray(pos_attr as GLuint);
+            gl::VertexAttribPointer(
+                pos_attr as GLuint,
+                3,
+                gl::FLOAT,
+                gl::FALSE,
+                3 * mem::size_of::<f32>() as i32,
+                ptr::null(),
+            );
+        }
+        BoxDrawer {
+            outline_program,
+            u_transform,
+            u_color,
+            vertex_array,
+            _buffer_position,
+            _buffer_indices,
+        }
+    }
+
+    // draws the unit box using any kind of transformation
+    fn draw_outlines_from_transformation(&self, matrix: &Matrix4<f32>, color: &Color) {
+        self.vertex_array.bind();
+
+        unsafe {
+            gl::UseProgram(self.outline_program.id);
+            gl::UniformMatrix4fv(self.u_transform, 1, false as GLboolean, matrix.as_ptr());
+            gl::Uniform4f(
+                self.u_color,
+                color.red,
+                color.green,
+                color.blue,
+                color.alpha,
+            );
+            gl::DrawElements(gl::LINES, 24, gl::UNSIGNED_INT, ptr::null());
+        }
+    }
+
+    // draws an outlined box around the given cube with the provided transformation matrix and color
```

Comment:
// Draws the outline of 'cube' using `color`. The cube is in world coordinates and we use 'world_to_gl_matrix' to transform it into screen space.

same comment about matrix being a type `world_to_gl` is probably sufficient.